Base/home/anon/sh/tests/valid.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/sh
+
+# Are comments ignored?
+# Sanity check: can we do && and || ?
+true || exit 2
+false && exit 2
+
+# Sanity check: can we pass arguments to 'test'?
+test yes = yes || exit 2
+
+# Sanity check: can we use $(command)?
+test $(echo yes) = yes || exit 2
```

Comment:
```suggestion
test "$(echo yes)" = yes || exit 2
```

Reply:
Also note that there's currently a bug in `test`: `test = yes` exits with 0.

So, without quotes, if `$(echo yes)` actually didn't work; `test = yes` would exit with `0` and this check would be ineffective.

Reply:
That should be fine, there's no particular need to stringify that.
it returns a `(yes)`, which is spliced in to form `test yes = yes`, stringifying it would create the same thing.

Reply:
oh ffs, shellcheck shouldn't be run on our shell, we're not bash nor are we sh.

Reply:
> So, without quotes, if $(echo yes) actually didn't work; test = yes would exit with 0 and this check would be ineffective.

I see what you mean.
Thanks for catching that.

Reply:
?? why?

Reply:
`"$(whatever)"` is just the string literal, "`$(whatever)`", no?

Reply:
why?
why which?

Reply:
No? `'$(whatever)'` would be the string literal.
`"$(whatever)"` is the stringification of the result of `$(whatever)`.
string interpolation is too useful to not include.

Reply:
But you can do string interpolation with `foo$(whatever)bar`, or if you need quotes, `"foo"$(whatever)"bar"` (though I'm, uh, not sure that would work, because `$()` gives you back what? — a list?)

So you want to make `""`s a stringification operator?

Reply:
`$()` gives you back a list, yes.
I want `""` to be a stringification op...yeah, that's about right.
though I consider it more like ruby's `""`, "a string that allows for interpolation".

Reply:
and `foo$(whatever)bar` would expand to many args if the `$()` returns a list of size > 1, and no args if it returns a list of size == 0, so it's not particularly...safe for string interpolation.

Reply:
Well yes, that's why it's problematic. I wouldn't object to either stringification or interpolation, but let's talk about what it actually does.

Is it actually a stringification operator? If so, then how do commands get stringified? Do lists get stringified as their individual items joined by spaces? Are spaces inside items escaped?

Reply:
The `echo` command just prints whitespace.

`test "$(echo)" = yes` will expand to `test "" = yes` -> `argv[0]="test" argv[1]="" argv[2]="=" argv[3]="yes"`

`test $(echo) = yes` will expand to `test  = yes` -> `argv[0]="test" argv[1]="=" argv[2]="yes"`

Reply:
> `test $(echo) = yes` will expand to `test = yes`

That depends on how we specify what `$()` does on empty input — does it evaluate to an empty list, or to a list with an empty element.

Reply:
Stringifying commands is a runtime error (though that's not implemented yet, so it just returns an empty list, which is then converted to an empty string).

lists are simply joined by spaces, no escaping is done, and that's probably for the best.

Reply:
> That depends on how we specify what $() does on empty input — does it evaluate to an empty list, or to a list with an empty element.

`$()` is currently defined as "take command output, split it by the value of `$IFS` (or `\n` if that's not defined), then spit the result back out as a list.
So it would return `()` on `$(echo)`, unless `IFS != "\n"`.

Reply:
But should it set `keep_empty` when splitting?..

Reply:
_that_, I am not sure of.
its current behaviour is, of course, not setting that, but I'm not sure about the implications of having it set.
I don't think it needs to keep empty segments, but I have no particular reason for that.

Reply:
Imagine a program like `find` that outputs a list of files. Say it outputs an empty line among them, to signify the current directory (instead of `.` as it's done usually). I think that empty line should make it to the list. But we don't want the extra blank line at the end to appear in the list, do we? E.g. `foo\nbar\n\nbaz\n` → `["foo", "bar", "", "baz"]`

Reply:
That hypothetical program really needs fixing :smile: 
hmm...but on the other hand, it's also useful to have it ignore the blanks...
perhaps we could simply add a variable that defines this behaviour? although it would still not do what you want, it'd either throw both empty lines out, or throw neither out...

Reply:
(hopefully) resolved by `setopt --inline_exec_keep_empty_segments`.
since that's a pain to write, try `setopt --in<tab>` :^)